pageserver/src/http/routes.rs

```diff
@@ -1091,6 +1175,64 @@ async fn post_tracing_event_handler(mut r: Request<Body>) -> Result<Response<Bod
     json_response(StatusCode::OK, ())
 }
 
+/// Common functionality of all the HTTP API handlers.
+///
+/// - Adds a tracing span to each request (by `request_span`)
+/// - Logs the request depending on the request method (by `request_span`)
+/// - Logs the response if it was not successful (by `request_span`
+/// - Shields the handler function from async cancellations. Hyper can drop the handler
+///   Future if the connection to the client is lost, but most of the pageserver code is
+///   not async cancellation safe. This converts the dropped future into a graceful cancellation
+///   request with a CancellationToken.
+async fn api_handler<R, H>(request: Request<Body>, handler: H) -> Result<Response<Body>, ApiError>
+where
+    R: std::future::Future<Output = Result<Response<Body>, ApiError>> + Send + 'static,
+    H: FnOnce(Request<Body>, CancellationToken) -> R + Send + Sync + 'static,
+{
+    // Spawn a new task to handle the request, to protect the handler from unexpected
+    // async cancellations. Most pageserver functions are not async cancellation safe.
+    // We arm a drop-guard, so that if Hyper drops the Future, we signal the task
+    // with the cancellation token.
+    let token = CancellationToken::new();
+    let guard = token.clone().drop_guard();
+    let handle = tokio::spawn(request_span(request, move |r| handler(r, token)));
+
+    let result = handle.await;
+
+    guard.disarm();
```

Comment:
I think we should retain the request cancelled logging via dropguard. It is extra information for debugging and we don't control the requester/timings/network.

I am also a concerned about removing that functionality from `request_span` because it will be used by safekeepers after #4119.

Reply:
Ok, I put that back.

Reply:
Hmm it's gone?

Reply:
Aah, it must be gone from the diffs and still being used from the request_span, so this is unchanged. Resolving again.